### PR TITLE
Fix Structure mode migration

### LINKED ISSUE #4411: [3.1.3] AttributeError: 'User' object has no attribute 'has_perm'
Hi,

Using orm['auth.Model'] in `0075_use_structure.py` migration can not have `has_perm` attribute:

>    File "/home/dcordel/workspace/webu/projects/ffcriavs/python-env/local/lib/python2.7/site-packages/cms/south_migrations/0075_use_structure.py", line 39, in forwards
>      if user.has_perm("cms.change_page"):
>  AttributeError: 'User' object has no attribute 'has_perm'

Using "real" `User` model (like in 3.1.2) is not a good solution. What do you recognize for this ? I could write a PR but I'm not sure about which solution is the best : using real User model ? Permission queryset on user and user's groups ? other ?

### PR DESCRIPTION
**Pull request by yakky**
Fix #4411 (hopefully)

### CODE REVIEW

File: cms/south_migrations/0075_use_structure.py [hunk 1]
```diff
@@ -32,12 +32,12 @@ def forwards(self, orm):
         try:
             ph_ctype = ContentType.objects.get(app_label=ph_model._meta.app_label, model=ph_model._meta.model_name)
             page_ctype = ContentType.objects.get(app_label=page_model._meta.app_label, model=page_model._meta.model_name)
-            permission, _ = Permission.objects.get_or_create(
+            permission, __ = Permission.objects.get_or_create(
                 codename='use_structure', content_type=ph_ctype, name=u"Can use Structure mode")
-            page_permission = Permission.objects.get(codename='change_page', content_type=page_ctype)
+            page_permission, __ = Permission.objects.get(codename='change_page', content_type=page_ctype)
```
Comment (DylannCordel) [inline cms/south_migrations/0075_use_structure.py]: should be `get_or_create` or remove the `__` var

Comment (yakky) [inline cms/south_migrations/0075_use_structure.py]: Done, thanks

**Comment by DylannCordel** [comment]
Hi @yakky 

If we use a custom permission backend, it could not work because permissions won't be tested "as usual". We should add a Warning for that cases. Or we could use the `django.contrib.auth.models._user_has_perm` function (but it uses real Models and not the old ones).

Instead this special case and my previous comment on `get_or_create`, LGTM according to the ["custom user requirements"](https://django-cms.readthedocs.org/en/develop/reference/configuration.html?highlight=abstractuser#custom-user-requirements) section of django CMS doc..

**Comment by vstoykov** [comment]
I tried this pull request without any custom backends or custom user model and forwards migrations from empty DB are ok. The only problem are backward migrations of 0010 and 0005 but I think that this is not so big problem.

Error for backward migrating 0010 is:

    ValueError: Cannot query "cms | placeholder | Can use Structure mode": Must be "Permission" instance.

And error for backward migrating 0005 is:

    django.db.migrations.migration.IrreversibleError: Operation <RunPython <function move_to_mp at 0x7f0bb558a7b8>> in cms.0005_auto_20140924_1039 is not reversible

**Comment by vstoykov** [comment]
Probably is a good idea for new installations to use squashed migration, because these datamigrations are unnecessary for new installations and will make maitanance for future release harder. https://docs.djangoproject.com/en/1.8/topics/migrations/#squashing-migrations

**Comment by yakky** [comment]
@DylannCordel @vstoykov do you agree to just catch all the exception in forward migrations and emit the warning?
@vstoykov  I updated the PR fixing the first error in backward migrating 0010.
The second one is a standard error when a bacward migration is not possible, identical to the similar message that exists in south.
Squashing migrations can be a nice option for 3.2, thanks for the suggestion

**Comment by yakky** [comment]
The 'real one' is #4472 which is against the correct branch. I'll leave this open to keep the discussion intact

**Comment by DylannCordel** [comment]
@yakky good enough for a migration. LGTM.

**Comment by vstoykov** [comment]
My opinion for catching all errors is 50/50. For empty database there are no sense to break. Also there are no sense of warning at all.
Probably the situation where error can be raised is when custom user model is used which can be checked easily and warning to be printed in this situation.
One more thing. If forward migration raises an error for some reason I thing that there is a big chance in backward migration also error to be raised.

**Comment by yakky** [comment]
@vstoykov the meaning of migrations it's just an heads up for the developer. Everything will still work despite any migration failing, but a check of the state of "use structure" permission is required.

**Comment by vstoykov** [comment]
ok then there are no real reason for migrations to break if this particular migration is failing for some reason and warning can be printed

**Comment by yakky** [comment]
@vstoykov @DylannCordel So, the migration now catches any exception and returns a warning allowing migration in any case

**Comment by yakky** [comment]
@DylannCordel @vstoykov merging above PR, I'll be glad if you could give it a last round

**Comment by vstoykov** [comment]
For me It's ok. If @DylannCordel is ok you can merge it.

File: cms/migrations/0010_migrate_use_structure.py [hunk 2]
```diff
@@ -14,31 +15,33 @@ def forwards(apps, schema_editor):
     try:
         ph_ctype = ContentType.objects.get_for_model(ph_model)
         page_ctype = ContentType.objects.get_for_model(page_model)
-        permission, _ = Permission.objects.get_or_create(
+        permission, __ = Permission.objects.get_or_create(
             codename='use_structure', content_type=ph_ctype, name=u"Can use Structure mode")
```
Comment (DylannCordel) [inline cms/migrations/0010_migrate_use_structure.py]: As `__future__.unicode_literals` is imported, shouldn't `u'strings'` be replaced by `'strings'` ?
Else, LGTM to be merged.

Comment (yakky) [inline cms/migrations/0010_migrate_use_structure.py]: Removed, it slipped in, but it's not consisten with the rest of the codebase

**Pull request closed without merging**